---
title: "R Notebook"
output: html_notebook
---

##### GENE SET ENRICHMENT ANALYSIS ######

```{r}

# citations:

# Gautier, L., Cope, L., Bolstad, B. M., and Irizarry, R. A. 2004. affy---analysis of Affymetrix GeneChip data at the probe level. Bioinformatics 20, 3 (Feb. 2004), 307-315.

# Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth, G.K. (2015). limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Research 43(7), e47.
  
# Müller K, Wickham H (2023). _tibble: Simple Data Frames_. R package version 3.2.1, <https://CRAN.R-project.org/package=tibble>.
  
# H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.

# Wickham H, François R, Henry L, Müller K, Vaughan D (2023). _dplyr: A Grammar of Data Manipulation_. R package version 1.1.4, <https://CRAN.R-project.org/package=dplyr>.
  
# Slowikowski K (2024). _ggrepel: Automatically Position Non-Overlapping Text Labels with 'ggplot2'_. R package version 0.9.6, <https://CRAN.R-project.org/package=ggrepel>.
  
# Wickham H, Hester J, Bryan J (2024). _readr: Read Rectangular Text Data_. R package version 2.1.5, <https://CRAN.R-project.org/package=readr>.
  
# Kolde R (2019). _pheatmap: Pretty Heatmaps_. R package version 1.0.12, <https://CRAN.R-project.org/package=pheatmap>.

# Luo, W. and Brouwer C., Pathview: an R/Bioconductor package for pathway-based data integration and visualization. Bioinformatics, 2013, 29(14): 1830-1831, doi: 10.1093/bioinformatics/btt285
  
# S Xu, E Hu, Y Cai, Z Xie, X Luo, L Zhan, W Tang, Q Wang, B Liu, R Wang, W Xie, T Wu, L Xie, G Yu. Using clusterProfiler to characterize multiomics data. Nature Protocols. 2024, doi:10.1038/s41596-024-01020-z

# Yu G (2024). _enrichplot: Visualization of Functional Enrichment Result_. doi:10.18129/B9.bioc.enrichplot <https://doi.org/10.18129/B9.bioc.enrichplot>, R package version 1.24.4, <https://bioconductor.org/packages/enrichplot>.

# Jahn N (2023). _europepmc: R Interface to the Europe PubMed Central RESTful Web Service_. R package version 0.4.3, <https://CRAN.R-project.org/package=europepmc>.

# Campitelli E (2024). _ggnewscale: Multiple Fill and Colour Scales in 'ggplot2'_. R package version 0.5.0, <https://CRAN.R-project.org/package=ggnewscale>.


library(affy)
library(limma)
library(tibble)
library(ggplot2)
library(dplyr)
library(ggrepel)
library(readr)
library(pheatmap)
library(pathview)
library(clusterProfiler)
library(enrichplot)
library(europepmc)
library(ggnewscale)

```

```{r}
#organism <- "org.Hs.eg.db"
#library(organism, character.only = TRUE)
```

# Get the genesets for mapping
```{r}
library(msigdbr)
```

```{r}
H_gene_sets = msigdbr(category = "H")
H_gene_sets
```


```{r}
C2_gene_sets = msigdbr(category = "C2")
C2_gene_sets
```


```{r}
C4_gene_sets = msigdbr(category = "C4")
C4_gene_sets
```


```{r}
msigdb_gene_sets <- rbind(H_gene_sets, C2_gene_sets, C4_gene_sets)
msigdb_gene_sets
```


```{r}
map = msigdb_gene_sets[, c("gs_id", "human_gene_symbol")]
map
```



This code will create variables in your global environment with names based on the .tsv files in the directory specified by base_path. Please ensure that the file paths are correct and the files exist at those locations. If you’re still encountering issues, please provide more details about the problem.

```{r}
# Define the base file path
base_path <- "/Users/WCVan/Downloads/pdac_analysis_vangie/liver_met_basal_results/filtered-degs/"

# Get the list of all .tsv files in the directory
file_names <- list.files(path = base_path, pattern = "\\.tsv$")

listofdataframes <- as.list(file_names)

# Read the files and assign to variables named after the file names
for (file_name in file_names) {
  # Remove the extension (.tsv) from the file name to use as the variable name
  var_name <- sub("\\.tsv$", "", file_name)
  
  # Read the file
  data <- read.csv(paste0(base_path, file_name), sep = "\t", header = TRUE)
  
  # Assign the data to a variable with the name stored in var_name
  assign(var_name, data, envir = .GlobalEnv)
}
```

```{r}
listofdataframes
```


# All
```{r}
df <- liver_met_basal_allgenes
```

```{r}
df
```

```{r}
df <- df[!apply(is.na(df) | df$ORF == "", 1, all),]
df$ORF <- sub("/.*", "", df$ORF)
df$ORF <- df$ORF
```

```{r}
df
```



```{r}
upreg <- df[df$logFC >= 1, ]
downreg <- df[df$logFC <= -1, ]
```


```{r}
upreg_gene_list <- upreg$ORF
downreg_gene_list <- downreg$ORF
```

```{r}
upreg_gene_list
```


```{r}
# we want the log2 fold change
original_gene_list <- df$logFC

# name the vector
names(original_gene_list) <- df$ORF

# omit any NA values
gene_list <- na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
gene_list <- sort(gene_list, decreasing = TRUE)
```

```{r}
gene_list
```

```{r}
df
```

```{r}
filtdf <- df %>% distinct(ORF, .keep_all = TRUE)
```

```{r}
df <- filtdf
```

```{r}
df
```

```{r}
gse <- GSEA(geneList = gene_list,
             minGSSize = 10,
             maxGSSize = 500,
             pvalueCutoff = 0.05,
             verbose = TRUE,
             pAdjustMethod = "fdr",
             gson = NULL,
             TERM2GENE = map,
             TERM2NAME = NA)
```

```{r}
genesets <- gse[]
```

```{r}
genesets
```

```{r}
#genesets$Term <- genesets$ID
#genesets$Genes <- genesets$core_enrichment
#genesets$FDR <- genesets$p.adjust
#genesets$Genes <- gsub("/", ", ", genesets$Genes)
```


```{r}
write_tsv(genesets, file = "/Users/WCVan/Downloads/pdac_analysis_vangie/liver_met_basal_results/clusterprofilerout_msigdb/GENESET_liver_met_basal_allgenes.tsv")
write_tsv(df, file = "/Users/WCVan/Downloads/pdac_analysis_vangie/liver_met_basal_results/clusterprofilerout_msigdb/liver_met_basal_allgenes.tsv")
```




# DAGs

```{r}
df <- liver_met_basal_met_specific_DAGs
```

```{r}
df
```

```{r}
df <- df[!apply(is.na(df) | df$ORF == "", 1, all),]
df$ORF <- sub("/.*", "", df$ORF)
df$ORF <- df$ORF
```

```{r}
df
```


```{r}
upreg <- df[df$logFC >= 1, ]
downreg <- df[df$logFC <= -1, ]
```


```{r}
upreg_gene_list <- upreg$ORF
downreg_gene_list <- downreg$ORF
```

```{r}
upreg_gene_list
```


```{r}
# we want the log2 fold change
original_gene_list <- df$logFC

# name the vector
names(original_gene_list) <- df$ORF

# omit any NA values
gene_list <- na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
gene_list <- sort(gene_list, decreasing = TRUE)
```

```{r}
gene_list
```

```{r}
df
```

```{r}
filtdf <- df %>% distinct(ORF, .keep_all = TRUE)
```

```{r}
df <- filtdf
```

```{r}
df
```

```{r}
gse <- GSEA(geneList = gene_list,
             minGSSize = 10,
             maxGSSize = 500,
             pvalueCutoff = 0.05,
             verbose = TRUE,
             pAdjustMethod = "fdr",
             gson = NULL,
             TERM2GENE = map,
             TERM2NAME = NA)
```

```{r}
genesets <- gse[]
```


```{r}
genesets
```

```{r}
#genesets$Term <- genesets$ID
#genesets$Genes <- genesets$core_enrichment
#genesets$FDR <- genesets$p.adjust
#genesets$Genes <- gsub("/", ", ", genesets$Genes)
```

```{r}
write_tsv(genesets, file = "/Users/WCVan/Downloads/pdac_analysis_vangie/liver_met_basal_results/clusterprofilerout_msigdb/GENESET_liver_met_basal_DAGs.tsv")
write_tsv(df, file = "/Users/WCVan/Downloads/pdac_analysis_vangie/liver_met_basal_results/clusterprofilerout_msigdb/liver_met_basal_DAGs.tsv")
```


# NDAGs

```{r}
df <- liver_met_basal_NDAGs
```

```{r}
df
```

```{r}
df <- df[!apply(is.na(df) | df$ORF == "", 1, all),]
df$ORF <- sub("/.*", "", df$ORF)
df$ORF <- df$ORF
```

```{r}
df
```

```{r}
upreg <- df[df$logFC >= 1, ]
downreg <- df[df$logFC <= -1, ]
```


```{r}
upreg_gene_list <- upreg$ORF
downreg_gene_list <- downreg$ORF
```

```{r}
upreg_gene_list
```


```{r}
# we want the log2 fold change
original_gene_list <- df$logFC

# name the vector
names(original_gene_list) <- df$ORF

# omit any NA values
gene_list <- na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
gene_list <- sort(gene_list, decreasing = TRUE)
```

```{r}
gene_list
```

```{r}
df
```

```{r}
filtdf <- df %>% distinct(ORF, .keep_all = TRUE)
```

```{r}
df <- filtdf
```

```{r}
df
```


```{r}
gse <- GSEA(geneList = gene_list,
             minGSSize = 10,
             maxGSSize = 500,
             pvalueCutoff = 0.05,
             verbose = TRUE,
             pAdjustMethod = "none",
             gson = NULL,
             TERM2GENE = map,
             TERM2NAME = NA)
```

```{r}
genesets <- gse[]
```


```{r}
genesets
```


```{r}
#genesets$Term <- genesets$ID
#genesets$Genes <- genesets$core_enrichment
#genesets$FDR <- genesets$p.adjust
#genesets$Genes <- gsub("/", ", ", genesets$Genes)
```


```{r}
write_tsv(genesets, file = "/Users/WCVan/Downloads/pdac_analysis_vangie/liver_met_basal_results/clusterprofilerout_msigdb/GENESET_liver_met_basal_NDAGs.tsv")
write_tsv(df, file = "/Users/WCVan/Downloads/pdac_analysis_vangie/liver_met_basal_results/clusterprofilerout_msigdb/liver_met_basal_NDAGs.tsv")
```

